File: Java/VendingMachine/src/MainFrame.java
import domain.Product;

import javax.swing.*;
import java.awt.*;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.util.List;

public class MainFrame extends JFrame {
    private JList<String> productList;
    private JTextField moneyField;
    private JButton buyButton;
    private DefaultListModel<String> listModel;

    public MainFrame(List<Product> products) {
        setTitle("Vending Machine");
        setSize(400, 300);
        setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);

        // Создание списка продуктов
        listModel = new DefaultListModel<>();
        for (Product product : products) {
            listModel.addElement(product.getName() + " - $" + product.getPrice());
        }
        productList = new JList<>(listModel);

        // Создание поля для ввода наличности
        moneyField = new JTextField(10);

        // Создание кнопки "купить товар"
        buyButton = new JButton("Купить товар");

        // Расположение компонентов на интерфейсе
        setLayout(new BorderLayout());
        add(new JScrollPane(productList), BorderLayout.CENTER);
        JPanel bottomPanel = new JPanel();
        bottomPanel.add(new JLabel("Введите сумму:"));
        bottomPanel.add(moneyField);
        bottomPanel.add(buyButton);
        add(bottomPanel, BorderLayout.SOUTH);

        // Обработка нажатия на кнопку "купить товар"
        buyButton.addActionListener(new ActionListener() {
            @Override
            public void actionPerformed(ActionEvent e) {
                int selectedIndex = productList.getSelectedIndex();
                if (selectedIndex != -1) {
                    String selectedItem = listModel.getElementAt(selectedIndex);
                    // Разбиваем строку, чтобы получить имя продукта и его цену
                    String[] parts = selectedItem.split(" - \\$");
                    String productName = parts[0];
                    double productPrice = Double.parseDouble(parts[1]);

                    // Получаем введенную сумму из поля для ввода
                    double enteredMoney = Double.parseDouble(moneyField.getText());

                    // Проверяем, хватает ли денег для покупки товара
                    if (enteredMoney >= productPrice) {
                        JOptionPane.showMessageDialog(MainFrame.this, "Вы купили " + productName);
                        // Обновляем баланс пользователя после покупки
                        moneyField.setText(String.valueOf(enteredMoney - productPrice));
                    } else {
                        JOptionPane.showMessageDialog(MainFrame.this, "Недостаточно средств");
                    }
                } else {
                    JOptionPane.showMessageDialog(MainFrame.this, "Выберите товар для покупки");
                }
            }
        });

        setVisible(true);
    }
}

File: Python/VendingMachine/main_frame.py
from PyQt5.QtWidgets import QApplication, QMainWindow, QListWidget, QLineEdit, QPushButton, QVBoxLayout, QMessageBox, \
    QWidget


class MainFrame(QMainWindow):
    def __init__(self, products):
        super().__init__()

        self.setWindowTitle("Vending Machine")
        self.setFixedSize(400, 300)

        # Создание списка продуктов
        self.product_list = QListWidget()
        for product in products:
            self.product_list.addItem(f"{product.name} - ${product.price}")

        # Создание поля для ввода наличности
        self.money_field = QLineEdit()

        # Создание кнопки "купить товар"
        self.buy_button = QPushButton("Купить товар")

        # Расположение компонентов на интерфейсе
        main_layout = QVBoxLayout()
        main_layout.addWidget(self.product_list)
        main_layout.addWidget(self.money_field)
        main_layout.addWidget(self.buy_button)

        central_widget = QWidget()
        central_widget.setLayout(main_layout)
        self.setCentralWidget(central_widget)

        # Обработка нажатия на кнопку "купить товар"
        self.buy_button.clicked.connect(self.on_buy_button_clicked)

    def on_buy_button_clicked(self):
        selected_index = self.product_list.currentRow()
        if selected_index != -1:
            selected_item = self.product_list.currentItem().text()
            # Разбиваем строку, чтобы получить имя продукта и его цену
            product_name, product_price = selected_item.split(" - ")
            product_price = float(product_price)

            # Получаем введенную сумму из поля для ввода
            entered_money = float(self.money_field.text())

            # Проверяем, хватает ли денег для покупки товара
            if entered_money >= product_price:
                QMessageBox.information(self, "Покупка", f"Вы купили {product_name}")
                # Обновляем баланс пользователя после покупки
                self.money_field.setText(f"{entered_money - product_price}")
            else:
                QMessageBox.warning(self, "Ошибка", "Недостаточно средств")
        else:
            QMessageBox.warning(self, "Ошибка", "Выберите товар для покупки")
